src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+        :type index_strings: List of str.
+        :param firstprivate_list: The list of firstprivate variables in the \
+                                  task region.
+        :type firstprivate_list: List of str.
+        :param private_list: The list of private variables in the task region.
+        :type private_list: List of str.
+        :param parallel_private: The list of private variables in the parent \
+                                 parallel region.
+        :type parallel_private: List of str.
+        :param start_val: Start/init value of outermost Loop contained in the \
+                          task region.
+        :type start_val: ???
+        :param stop_val: Stop value of outermost Loop contained in the task \
+                         region.
+        :type stop_val: ???
+
+        :raises GenerationError: If node.operator is not ADD or SUB
+        :raises GenerationError: If the children of node do not contain one \
+                                 Reference and one Literal
+        :raises GenerationError: If the child Reference is a private variable.
+        :raises GenerationError: If the child Reference is a shared variable.
+        '''
+        from psyclone.psyir.backend.fortran import FortranWriter
+        writer = FortranWriter()
+        # Binary Operation check
+        if node.operator is not \
+           BinaryOperation.ADD and \
+           node.operator is not \
+           BinaryOperation.SUB:
+            raise GenerationError(
+                "Binary Operator of type {0} used "
+                "as in index inside an "
+                "OMPTaskDirective which is not "
+                "supported".format(
+                    node.operator))
+        # We have ADD or SUB BinaryOperation
+        if not( (type(node.children[0]) is \
+           Reference and type(node.children[1]) is\
+           Literal) or (type(node.children[0])\
+           is Literal and type(node.children[1])\
+           is Reference)):
+            raise GenerationError(
+                "Children of BinaryOperation are of "
+                "types {0} and {1}, expected one "
+                "Reference and one Literal when"
+                " used as an index inside an "
+                "OMPTaskDirective.".format(
+                    type(node.children[0]).
+                    __name__, type(node.children[1]).
+                __name__))
+
+        # Have Reference +/- Literal, analyse
+        # and create clause appropriately
+        index_name = "" 
+        if type(node.children[0]) is Reference:
+            index_name = writer(node.children[0].symbol)
+        if type(node.children[1]) is Reference:
+            index_name = writer(node.children[1].symbol)
+        index_private = (index_name in 
+                         parallel_private)
+        if index_private:
+            if index_name in private_list:
+                # FIXME I think this should be
```

Comment:
Just in case you're not aware (and it helps), the dependency analysis now makes use of SymPy.